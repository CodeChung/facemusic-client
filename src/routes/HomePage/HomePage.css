.camera {
    position: relative;
    justify-content: center;
    width: 80%;
    margin: 0 auto;
}

.camera-page, .calendar-page {
    background: #F1FFFA;
}

.current-img {
    width: 768px;
    height: auto;
    position: relative;
    margin: 0 auto;
}

.current-img img {
    width: 100%;
    height: auto;
    /* position: absolute; */
    left: 0;
    top: 0;
    z-index: 1;
}

.current-img button {
    font-size: 20px;
    padding: 15px;
    border-radius: 2.5px;
    background-color: white;
    box-shadow: 0 !important;
}

.current-img button:hover {
    background: black;
    border: white;
    color: white;
    outline: none;
}

.current-img .analyze-img {
    position: absolute;
    left: 20px;
    bottom: 10px;
    z-index: 100;
}

.current-img .delete-img {
    position: absolute;
    right: 20px;
    bottom: 10px;
    z-index: 100;
}

.emotion-graph {
    border-radius: 50px;
}

.emotion-graph img {
    margin-bottom: 30px;
}

.emotion-graph form {
    padding: 15px;
}

.emotion-graph textarea {
    width: 50%;
    border-radius: 5px;
    height: 30vh;
    padding: 5px;
}

.emotion-graph textarea:focus {
    outline: none;
}

.hidden {
    visibility: hidden;
    height: 576px;
    padding: 5px;
}

.photo-booth video {
    border-radius: 10px;
}

.recommendation-list {
    display: flex;
    flex-flow: wrap;
    margin: 0 auto;
    justify-content: center;
}

a {
    text-decoration: none;
    color: black;
}

.spinner {
    display: flex;
    justify-content: center;
    align-items: center;
}

.warning {
    z-index: 100;
}

@media(max-width:768px){
    .current-img {
        width: 100%;
    }
    .emotion-graph img {
        width: 80%;
    }
}
