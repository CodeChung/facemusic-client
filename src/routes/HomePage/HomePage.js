import React from 'react';
import { Redirect } from 'react-router-dom';
import Camera from 'react-html5-camera-photo';
import 'react-html5-camera-photo/build/css/index.css';
import Results from '../../components/Results/Results'
import './HomePage.css'
import ServerApiService from '../../services/server-api-service';
import { MoonLoader } from 'react-spinners';
import Entry from '../../components/Entry/Entry';
import AppContext from '../../AppContext';

class HomePage extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            photo: '',
            error: '',
            analyzedPhoto: '',
            emotion: {},
            entry: {},
            artistSeeds: [],
            trackSeeds: [],
            artists: [],
            tracks: [],
            loading: false,
            loaded: false,
        }
    }
    componentDidMount() {
        ServerApiService.getSavedSeeds()
          .then(res => {
            const { tracks, artists } = res
            this.setState({ tracks, artists, loaded: true })
          })
    }
    onTakePhoto(dataUri) {
        this.setState({photo: dataUri})
    }
    deletePhoto() {
        this.setState({
            photo: '',
            error: '',
        })
    }
    analyzePhoto() {
        this.setState({ loading: true })
        const photoData = this.state.photo
        const body = JSON.stringify({img: photoData})
        ServerApiService.convertPhotoToEmotion(body)
            .then(res => {
                const emotionData = res.faceAttributes.emotion
                const neutralVal = emotionData.neutral
                const emotion = {}
                for (const [key, val] of Object.entries(emotionData)) {
                    if (key !== 'neutral') {
                        emotion[key] = val / (1 - neutralVal)
                    }
                }
                this.setState({
                    error: '',
                    emotion,
                    analyzedPhoto: res.url,
                    loading: false,
                })
            })
            .catch(res => {
                this.setState({ error: res.error, loading: false })
            })
    }
    //either display webcam or the photo that was taken
    renderPhotoDisplay() {
        const { artists, loading, tracks } = this.state
        const photo = this.state.photo
        if (artists.length === 0 && tracks.length === 0) {
            return <Redirect to='/seeds' />
        } else {
            return !photo ?
            <div className='home-photo photo-booth'>
                <Camera
                    onTakePhoto={(dataUri) => this.onTakePhoto(dataUri)}
                />
            </div>
            :
            <div className='home-photo current-img'>
                <img
                    width={768}
                    src={photo}
                    alt='current'/>
                <button className='analyze-img' onClick={() => this.analyzePhoto()}>Analyze</button>
                <button className='delete-img' onClick={() => this.deletePhoto()}>Delete</button>
                {loading && <MoonLoader />}
            </div>
        }
    }
    renderView(context) {
        if (Object.keys({}) > 0) {
            return <Entry />
        } else {
            return Object.keys(this.state.emotion).length > 0 ?
            <Results 
                emotion={this.state.emotion} 
                photo={this.state.analyzedPhoto}
                />
            :
            <div className="camera">
                <h2>Today's Mood</h2>
                {this.renderPhotoDisplay()}
            </div>
        }
    }
    render() {
        const context = this.context
        if (!this.state.loaded) {
            return (
                <section className='spinner'>
                    <MoonLoader />
                </section>
            )
        }
        return (
            <section className='camera-page'>
                {this.renderView(context)}
            </section>
            
        )
    }
}

HomePage.contextType = AppContext

export default HomePage