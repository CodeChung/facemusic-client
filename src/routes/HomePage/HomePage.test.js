import React from 'react';
import ReactDOM from 'react-dom';
import HomePage from './HomePage';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';



describe(`HomePage Component`, () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<HomePage />, div);
    ReactDOM.unmountComponentAtNode(div);
  });
  
  it('renders loading given no props', () => {
    const wrapper = shallow(<HomePage />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('renders camera after componentDidMount', () => {
    const wrapper = mount(<HomePage />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
}) 
