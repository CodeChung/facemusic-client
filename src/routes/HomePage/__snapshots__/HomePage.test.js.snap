// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage Component renders camera after componentDidMount 1`] = `
<HomePage>
  <section
    className="spinner"
  >
    <Loader
      color="#000000"
      css={Object {}}
      loading={true}
      size={60}
      sizeUnit="px"
    >
      <EmotionCssPropInternal
        __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
        css={
          Array [
            Object {
              "map": undefined,
              "name": "so5deu",
              "next": Object {
                "name": "animation-12jek8e",
                "next": undefined,
                "styles": "@keyframes animation-12jek8e{
  100% {transform: rotate(360deg)}
}",
              },
              "styles": "
      position: relative;
      width: 77.14285714285714px;
      height: 77.14285714285714px;
      animation: animation-12jek8e 0.6s 0s infinite linear;
      animation-fill-mode: forwards;
    ",
            },
            Object {},
          ]
        }
      >
        <div
          className="css-so5deu"
        >
          <EmotionCssPropInternal
            __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
            css={
              Object {
                "map": undefined,
                "name": "1nbccn9",
                "next": Object {
                  "name": "animation-12jek8e",
                  "next": undefined,
                  "styles": "@keyframes animation-12jek8e{
  100% {transform: rotate(360deg)}
}",
                },
                "styles": "
      
      width: 8.571428571428571px;
      height: 8.571428571428571px;
      border-radius: 100%;
    ;
      background-color: #000000;
      opacity: 0.8;
      position: absolute;
      top: 25.714285714285715px;
      animation: animation-12jek8e 0.6s 0s infinite linear;
      animation-fill-mode: forwards;
    ",
              }
            }
          >
            <div
              className="css-1nbccn9"
            />
          </EmotionCssPropInternal>
          <EmotionCssPropInternal
            __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
            css={
              Object {
                "map": undefined,
                "name": "5m0nb3",
                "next": undefined,
                "styles": "
      
      width: 60px;
      height: 60px;
      border-radius: 100%;
    ;
      border: 8.571428571428571px solid #000000;
      opacity: 0.1;
    ",
              }
            }
          >
            <div
              className="css-5m0nb3"
            />
          </EmotionCssPropInternal>
        </div>
      </EmotionCssPropInternal>
    </Loader>
  </section>
</HomePage>
`;

exports[`HomePage Component renders loading given no props 1`] = `
<section
  className="spinner"
>
  <Loader
    color="#000000"
    css={Object {}}
    loading={true}
    size={60}
    sizeUnit="px"
  />
</section>
`;
